### PR TITLE
ref(integrations): Change VSTS display name.

### PR DESCRIPTION
**Pull request by lauryndbrown**
Dependent on #9433. We want to change VSTS issue names to display as `<project_name>#<issue_id>`, but this convention has no meaning in VSTS. This change uses a newly created field to store the `display_name` in the `ExternalIssue` model, and uses the frontend to display it to the user.

Does not currently handle default value changes.

### CODE REVIEW

File: src/sentry/integrations/vsts/issues.py [hunk 1]
```diff
@@ -31,14 +31,14 @@ def get_create_issue_config(self, group, **kwargs):
         project_choices = []
         initial_project = ('', '')
         for project in projects:
-            project_choices.append((project['id'], project['name']))
-            if project['id'] == self.default_project:
+            project_id_and_name = '%s#%s' % (project['id'], project['name'])
+            project_choices.append((project_id_and_name, project['name']))
+            if project_id_and_name == self.default_project:
```
Comment (lauryndbrown) [inline src/sentry/integrations/vsts/issues.py]: I'm not certain this works... I need to figure out how the default values work.

File: src/sentry/integrations/issues.py [hunk 2]
```diff
@@ -146,6 +146,17 @@ def make_external_key(self, data):
         """
         return data['key']
 
+    def get_issue_display_name(self, external_issue_key):
+        """
+
```
Comment (EvanPurkhiser) [inline src/sentry/integrations/issues.py]: nit: Blank space

File: src/sentry/api/serializers/models/integration.py [hunk 3]
```diff
@@ -152,12 +152,15 @@ def get_attrs(self, item_list, user, *args, **kwargs):
         ints_by_id = {i.id: i for i in item_list}
         for ei in external_issues:
             # TODO(jess): move into an external issue serializer?
+            installation = ints_by_id[ei.integration_id].get_installation(
+                self.group.organization.id)
```
Comment (markstory) [inline src/sentry/api/serializers/models/integration.py]: Any chance of a KeyError here?

File: src/sentry/integrations/issues.py [hunk 4]
```diff
@@ -146,6 +146,15 @@ def make_external_key(self, data):
...
         return data['key']
 
+    def get_issue_display_name(self, external_issue_key):
+        """
+        Returns the display name of the issue.
+
+        This is not required but helpful for integrations whose external issue key
+        does not match the disired display name.
+        """
+        return None
```
Comment (markstory) [inline src/sentry/integrations/issues.py]: Should this be an empty string to maintain type consistency with other implementations of this method?

File: src/sentry/integrations/vsts/issues.py [hunk 5]
```diff
@@ -31,14 +31,14 @@ def get_create_issue_config(self, group, **kwargs):
         project_choices = []
         initial_project = ('', '')
         for project in projects:
-            project_choices.append((project['id'], project['name']))
-            if project['id'] == self.default_project:
+            project_id_and_name = '%s#%s' % (project['id'], project['name'])
```
Comment (markstory) [inline src/sentry/integrations/vsts/issues.py]: You append join these values together and then split them later on (line 72). Could you use a tuple/dict instead to avoid joining/splitting?

File: src/sentry/integrations/vsts/issues.py (hunk 1 above)
Comment (lauryndbrown) [reply src/sentry/integrations/vsts/issues.py]: Pulling out default project changes because they are not implemented yet

File: src/sentry/integrations/vsts/issues.py (hunk 5 above)
Comment (lauryndbrown) [reply src/sentry/integrations/vsts/issues.py]: @markstory is it possible to use a tuple for the id of a js choice field? Js is not my primary language by any means; I wasn't aware that was possible.

File: src/sentry/integrations/issues.py (hunk 4 above)
Comment (lauryndbrown) [reply src/sentry/integrations/issues.py]: Hmm... so another js question: does empty string evaluate to False in js? If so it sounds like a pretty easy change to me.

File: src/sentry/api/serializers/models/integration.py (hunk 3 above)
Comment (lauryndbrown) [reply src/sentry/api/serializers/models/integration.py]: It doesn't appear so. If you take a look at line 148, the `external_issues` are a list that have `integration_ids` in the `item_list` that's used to make the `ints_by_id` dict.

File: src/sentry/integrations/issues.py (hunk 4 above)
Comment (markstory) [reply src/sentry/integrations/issues.py]: Yes. Empty strings are [falsy in javascript](https://developer.mozilla.org/en-US/docs/Glossary/Falsy).

```js
> if ('') { console.log('truthy') } else { console.log('falsy;') }
falsy;
```

File: src/sentry/integrations/vsts/issues.py (hunk 5 above)
Comment (markstory) [reply src/sentry/integrations/vsts/issues.py]: No it wouldn't be. I didn't know this was ending up in a clientside picker. My bad.

File: src/sentry/integrations/vsts/issues.py [hunk 6]
```diff
@@ -31,14 +31,15 @@ def get_create_issue_config(self, group, **kwargs):
...
+            project_choices.append((project_id_and_name, project['name']))
+            # TODO(lb): Properly handle default project after it has been implemented.
+            if project_id_and_name == self.default_project:
                 initial_project = project['name']
         return [
             {
                 'name': 'project',
```
Comment (EvanPurkhiser) [inline src/sentry/integrations/vsts/issues.py]: This is an asside and not part of this PR, but shouldn't this be capital?

File: src/sentry/integrations/vsts/issues.py (hunk 5 above)
Comment (EvanPurkhiser) [reply src/sentry/integrations/vsts/issues.py]: I'm kind of confused why we need to combine project and name in general? is the ID not unique? I may be forgetting context here

**Review by evanpurkhiser — APPROVED** [review]
Just had that one question about why we need to combine the project ID and name, but other than that it looks good!

**Review by markstory — APPROVED** [review]
I am light on project context, but this looks good to me.

File: src/sentry/integrations/vsts/issues.py (hunk 5 above)
Comment (lauryndbrown) [reply src/sentry/integrations/vsts/issues.py]: @EvanPurkhiser Yes the id is unique, but I wanted the name so I didn't have to look it up again right after the user presses submit on the form. Otherwise it requires another api call. If there's a different way to do that do let me know! I can follow up with another pr and make improvements

File: src/sentry/integrations/vsts/issues.py (hunk 6 above)
Comment (lauryndbrown) [reply src/sentry/integrations/vsts/issues.py]: `'name'` you mean? I kind of assume the frontend takes care of how things are displayed and I'm just providing the content. But if they need to be capitalized on the backend do let me know.
